### DISCUSSION TITLE
Periodic BC for Discontinuous Galerkin

### DISCUSSION THREAD

**Discussion by DaijunHU** [idaholab/moose#18580] [Q&A General] [answered]
Dear MOOSE users,

We are implementing the DG kernels for advection, but for the moment, the periodic BC does not work.
An example input file is here https://github.com/ngrilli/c_pfor_am/blob/main/test/tests/DGKernels/PosNegStraightEdge.i

In our view the DG kernel works only in the internal elements but for the variables on the boundaries the neighbor parts (e.g. _u_neighbor) do not work well.

Has anyone implemented Periodic BC for some Discontinuous Galerkin applications?
Any suggestion is very much appreciated.

Best Regards,
Daijun

**Comment by ngrilli** [comment]
We have noticed that periodic boundary conditions are activated only when the shape functions are FIRST LAGRANGE
but they do not work when the shape functions are CONSTANT MONOMIAL or FIRST MONOMIAL.

Does anybody know the reason for that?
And what could be a solution to make PBC working with CONSTANT MONOMIAL?

Thank you very much in advance for your suggestions.

Nicolò

**Comment by YaqiWang** [comment, marked as answer]
Is CG having the strong continuity along the paired periodic boundaries? I am not sure if we can do that with DG. I think we should have something like a DG kernel to connect a paired sides. Do you get a run-time error for doing periodic boundary condition with DG? If not, the code at least should throw an error. Hope someone else can give you a better answer.

**Reply by DaijunHU** [reply]
Hi Yaqi, 
Thank you so much for your reply. 
Currently we do not have an error for doing periodic boudary condition with DG kernels, it just do not work at boudary elements but the simulations can move on. 
 I think Nicolò's answer above points to a possible solution  for this, we want to check if there is PBC that can work with the shape function  CONSTANT MONOMIAL we are using for our simulation.
Thank you!

Best regards,
Daijun

**Reply by ngrilli** [reply]
Dear @YaqiWang 

Yes, I believe CG (with first order Lagrange) with PBC removes the degrees of freedom of the variable along one of the paired surfaces, therefore there is continuity in the strong sense.
I am still not sure why this is not possible with constant monomial shape functions, it should be technically the same.
This does not depend specifically on the DGKernel, the same happens with CG, 
but you never use constant monomial in that case.

Yes, as you point out, we should develop a kernel such that the variable at the neighboring element: _u_neighbor
can be fetched in the elements in the connected pair side.
Is there a function in moose that can provide the neighboring element in a periodic boundary?

The current DGKernel is not called for the boundary faces, therefore adding periodic BC has no effect, 
the simulation can run though, but boundary condition remains zero flux.

Nicolò

**Reply by YaqiWang** [reply]
> Is there a function in moose that can provide the neighboring element in a periodic boundary?

libMesh probably has a function to allow us to find the periodic neighbor. But MOOSE probably does not have the support for properly evaluating coupled variables, material properties, etc. on the periodic neighbor. @lindsayad are these right?

**Reply by lindsayad** [reply]
Yes we have [`Elem::topological_neighbor`](https://mooseframework.inl.gov/docs/doxygen/libmesh/classlibMesh_1_1Elem.html#ae81b9875caed609205c970d9cc0134a2) for finding a periodic neighbor. @YaqiWang is right that all the plumbing in MOOSE is set up just to query and reinit finite element shape functions, variables, materials etc. for connected neighbors (neighbors accessible with `elem->neighbor_ptr`). It would take some changes to support periodic neighbors
